### PR TITLE
Handle null type from semanticModel

### PR DESCRIPTION
**Pull request by jmarolf**
Fixes https://github.com/dotnet/roslyn/issues/6682

@dotnet/roslyn-ide

### CODE REVIEW

**Comment by jmarolf** [comment]
retest vsi please

File: src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs [hunk 1]
```diff
@@ -642,26 +642,44 @@ private bool ReplacementBreaksQueryClause(QueryClauseSyntax originalClause, Quer
...
-            if (originalExpression == null || originalTargetType == null || newExpression == null || newTargetType == null)
+            Conversion? originalConversion = null;
+            Conversion? newConversion = null;
+
+            if (HasNonNullType(this.OriginalSemanticModel, originalExpression) &&
+                HasNonNullType(this.SpeculativeSemanticModel, newExpression))
             {
-                return false;
+                originalConversion = this.OriginalSemanticModel.ClassifyConversion(originalExpression, originalTargetType);
+                newConversion = this.SpeculativeSemanticModel.ClassifyConversion(newExpression, newTargetType);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs]: I find these 5 lines hard to read.  Why if it has a NonNullType can you then call ClassifyConversion on some other type?

File: src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs [hunk 2]
```diff
@@ -642,26 +642,44 @@ private bool ReplacementBreaksQueryClause(QueryClauseSyntax originalClause, Quer
...
+            }
+            else
+            {
+                ITypeSymbol originalConvertedTypeSymbol = null;
+                if (TryGetConvertedTypeForExpression(this.OriginalSemanticModel, originalExpression, out originalConvertedTypeSymbol))
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs]: Why not just get the converted type, and then check if it is null right here?

File: src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs [hunk 3]
```diff
@@ -642,26 +642,44 @@ private bool ReplacementBreaksQueryClause(QueryClauseSyntax originalClause, Quer
...
+                {
+                    originalConversion = this.OriginalSemanticModel.Compilation.ClassifyConversion(originalConvertedTypeSymbol, originalTargetType);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs]: All this code (along with the above code in the ```if``` portoin) seems ripe for some sort of cleanup.  Aren't you just basically saying:

```
var typeINfo = model.GetTypeInfo(expression)
var type = typeInfo.Type ?? typeInfo.ConvertedType
ver conversion = model.ClassifyConvertion(type, originalTargetType)
```

File: src/Workspaces/VisualBasic/Portable/Utilities/SpeculationAnalyzer.vb [hunk 4]
```diff
@@ -518,21 +518,47 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Utilities
...
         Protected Overrides Function ConversionsAreCompatible(originalExpression As ExpressionSyntax, originalTargetType As ITypeSymbol, newExpression As ExpressionSyntax, newTargetType As ITypeSymbol) As Boolean
-            If originalExpression Is Nothing OrElse originalTargetType Is Nothing OrElse newExpression Is Nothing OrElse newTargetType Is Nothing Then
-                Return False
-            End If
+            Dim originalConversion As Conversion?
+            Dim newConversion As Conversion?
 
-            Dim originalConversion = Me.OriginalSemanticModel.ClassifyConversion(originalExpression, originalTargetType)
-            Dim newConversion = Me.SpeculativeSemanticModel.ClassifyConversion(newExpression, newTargetType)
+            If HasNonNullType(Me.OriginalSemanticModel, originalExpression) AndAlso
```
Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/Utilities/SpeculationAnalyzer.vb]: This code appears identical to the above C# version.  Why do we have both?  Can we not share anything here?

Comment (jmarolf) [inline src/Workspaces/VisualBasic/Portable/Utilities/SpeculationAnalyzer.vb]: The main problem is that [Conversion](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Semantics/Conversions/Conversion.cs) has no common type between VB and C#.  So we can't call ClassifyConversion from an abstract method.  What type would ClassifyConversion return?

Comment (CyrusNajmabadi) [inline src/Workspaces/VisualBasic/Portable/Utilities/SpeculationAnalyzer.vb]: Be generic on some TConversion type.

File: src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs (hunk 3 above)
Comment (jmarolf) [inline src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs]: I'll try and clean it up.  

`this.OriginalSemanticModel.ClassifyConversion(originalExpression, originalTargetType)`

does not have the same semantics as 

```C#
var typeinfo = this.OriginalSemanticModel.GetTypeInfo(originalExpression)
var type = typeInfo.Type ?? typeInfo.ConvertedType
var conversion = model.ClassifyConvertion(type, originalTargetType)
```

The second case is a fallback for when we know that `this.OriginalSemanticModel.ClassifyConversion(originalExpression, originalTargetType)` will throw and error.

**Comment by jmarolf** [comment]
test vsi please

**Comment by dpoeschl** [comment]
FYI @tannergooding, http://dotnet-ci.cloudapp.net/job/dotnet_roslyn_prtest_win_tst_eta/74/console failure with no Failure Indication and no build artifacts / etc.

File: src/EditorFeatures/CSharpTest/Diagnostics/SimplifyTypeNames/SimplifyTypeNamesTests.cs [hunk 5]
```diff
@@ -3287,5 +3287,14 @@ public class Class1
     public int this[int index] => (int)0;
 }");
         }
+
+        [WorkItem(6682, "https://github.com/dotnet/roslyn/issues/6682")]
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsSimplifyTypeNames)]
+        public async Task TestThisWithNoType()
+        {
+            await TestAsync(
+@"class Program { dynamic x = 7 ; static void Main ( string [ ] args ) { [|this|] . x = default(dynamic) ; } } ",
```
Comment (dpoeschl) [inline src/EditorFeatures/CSharpTest/Diagnostics/SimplifyTypeNames/SimplifyTypeNamesTests.cs]: C# didn't actually have this bug, right? Just adding coverage?

Comment (jmarolf) [inline src/EditorFeatures/CSharpTest/Diagnostics/SimplifyTypeNames/SimplifyTypeNamesTests.cs]: Correct

File: src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs [hunk 6]
```diff
@@ -707,5 +706,9 @@ protected override bool IsReferenceConversion(Compilation compilation, ITypeSymb
         {
             return compilation.ClassifyConversion(sourceType, targetType).IsReference;
         }
+
+        protected override Conversion ClassifyConversion(SemanticModel model, ExpressionSyntax expression, ITypeSymbol targetType) => model.ClassifyConversion(expression, targetType);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/CSharp/Portable/Utilities/SpeculationAnalyzer.cs]: wrap

**Comment by CyrusNajmabadi** [comment]
:+1:

**Pull request merged**